README.md
# fnx_rest

Set of REST tools which work nicely with Dart 2 / Angular.

fnx_rest is oriented to be developer and Angular friendly and is particularly useful
when creating boring CRUD applications with many similar API calls. 

## Example

```dart
    import 'package:fnx_rest/fnx_rest_browser.dart';
//     import 'package:fnx_rest/fnx_rest_io.dart'; (when not in a browser)       

    RestClient root = BrowserRestClient.root("/api/v1");        
    RestResult response = await root.child("/users").get();
    List users = response.data;
```
    
## Angular support

You can define `root` REST client, add your API keys and other additional headers to it
and inject this root client with Angular's dependency injection to your elements and/or services.
    
```dart
    RestClient root = BrowserRestClient.root("/api/v1");            
          
    // your component
    class MyApp {
        RestClient restRoot;
        MyApp(this.restRoot);        
    }
    
    // add custom headers, for example after user's signing in
    restRoot.setHeader("Authorization", authKey);       
```

(see Angular docs for DI details)     
    
RestClient is hierarchical:
    
```dart
    RestClient root = BrowserRestClient.root("/api/v1");   //  /api/v1
    RestClient users = root.child("/users");            //  /api/v1/users            
    RestClient john = users.child("/123");              //  /api/v1/users/123
```
    
All children inherit configuration of their parents, but are allowed
to override it.

RestClient supports query parameters:
    
```dart
    RestClient limitedUsers = users.setParam('limit', '1000');  //  /api/v1/users?limit=1000            
```

Typically you would create a child of the root rest
client in your component like this:

```dart
    class UserManagement {
        RestClient users;
        UserManagement(RestClient root) {
            users = root.child("/users"); // endpoint /api/v1/users
        }
    }
```
    
Every instance of RestClient has bool `working` property, which indicates whether this client
is currently processing a request/response or not. You can use it to indicate "working"
state to the user:

    <p *ngIf="users.working">Sending user data to server ...</p>
    
This property is recursively propagated to client's parents so you can indicate
this "working" state on any level. Locally (for a form),
or globally (for the whole app).
     
    // update user     
    users.put(...);

Until the request is processed, `john.working == true`, `users.working == true` and `root.working == true`.

    // read users
    users.get( ... )

In this case `john.working == false` but `users.working == true` and `root.working == true`.

You can easily
use this behaviour to disable a form and all it's buttons after submitting edited
user data, but in the same time you can have universal
global indicator of any HTTP communication (in your app status bar, for example).

## HTTP methods

RestClient has following methods:

```dart
    Future<RestResult> get({dynamic data, Map<String, String> headers}) ...
    Future<RestResult> post(dynamic data, {Map<String, String> headers}) ...
    Future<RestResult> put(dynamic data, {Map<String, String> headers}) ...
    Future<RestResult> delete({dynamic data, Map<String, String> headers}) ...
    Future<RestResult> head({Map<String, String> headers}) ...
```
    
Use optional parameter `headers` to specify custom ad-hoc headers you need
in this call only. Headers will be merged with your `RestClient` default headers,
it's parent's headers etc. up to the root `RestClient`. 

Don't use this parameter to specify `Content-Type` or `Accept` headers, see below. 

## RestResult

Each call returns `Future<RestResult>`. RestResult contains `status` (HTTP status, int) 
and `data` which are already converted to your
desired type (see below) - Dart Map or Dart List by default. 

## Request/response serialization

By default, the root client is configured to produce and consume JSON and
Dart Maps and Lists.

You can easily customize this behaviour to accept or produce any binary data:
 
```dart
    RestClient img = root.child("/images");        //  /api/v1/images     
    img.acceptsBinary("image/png");
    img.producesBinary("image/png");
```
    
Such data will be sent and received as `List<int>`.
 
You can also inject any custom serialization or deserialization you need:

```dart
    /*
    typedef Serializer = dynamic Function(dynamic payload, Map<String, String> requestHeaders);
    typedef Deserializer = dynamic Function(Response response);
    */

    client.accepts("text/csv", myCsvDeserializer);
    client.produces("text/csv", myCsvSerializer);
```
    
This configuration is inherited by client's children.                         
                
## Work in progress
                
Suggestions, pull requests and bugreports are more than welcome.                

CHANGELOG.md
# Changelog

## 4.1.1

Added optional "HttpClient? inner" parametr for IoRestClient, allowing user to specify details of the http client (mainly proxy, but possibly low-level TCP/IP settings etc).

## 4.1.0

Added PATCH method. *(facepalm)*

## 4.0.1

Null safe.

## 4.0.0

Maintenance release, mental preparation for null safe jump.


## 3.5.0

* pedantic linter

## 3.2.0

* RestResult now allows you to access response headers.

## 3.1.0

* serializers can now modify outgoing headers
* HTTP methods (get, post, ...) never throw a HttpException,
  not even with 500 http status response
* method RestResult.assertSuccess() when you are not interested in response body (successData)

## 3.0.0

* More flexible serializers and deserializers. (_Breaking api change if you implement your own custom serializers or deserializers_)
* Added `urlWithParams` getter for simpler access to url rendered with parameters
* Default json de/serializers now not throw on whitespace string
* Updated readme

## 2.2.0

Newer version of http client.
Example.

## 2.1.0

Added support for streaming requests.

## 0.0.1

- Initial version


test/debouncer_test.dart
import 'dart:async';
import 'package:fnx_rest/src/debouncer.dart';
import 'package:test/test.dart';

void main() {
  group("Debouncer", () {
    test("delivers only the last value", () async {
      StreamController<int> ctrl = new StreamController<int>(sync: true);
      Stream<int> stream = ctrl.stream;
      Debouncer<int, int> debouncer = new Debouncer(new Duration(milliseconds: 100), sync: true);
      Stream<int> t = stream.transform(debouncer);
      ctrl.add(1);
      ctrl.add(2);
      ctrl.add(3);
      int result = await t.first;
      expect(result, equals(3));
    });

    test("waits at least specified amount of time before it will deliver next value", () async {
      StreamController<int> ctrl = new StreamController<int>(sync: true);
      Stream<int> stream = ctrl.stream;
      Duration duration = new Duration(milliseconds: 100);
      Debouncer<int, int> debouncer = new Debouncer(duration, sync: true);
      DateTime start = new DateTime.now();
      Stream<int> t = stream.transform(debouncer);
      ctrl.add(1);
      await t.first;

      DateTime finished = new DateTime.now();

      DateTime shouldTakeAtLeast = start.add(duration);

      int diff = finished.millisecondsSinceEpoch - shouldTakeAtLeast.millisecondsSinceEpoch;

      expect(diff, greaterThanOrEqualTo(0));
    });
  });
}


test/io_client_test.dart
import 'dart:io';

import 'package:fnx_rest/fnx_rest_io.dart';
import 'package:test/test.dart';

void main() {
  group('IoRestClient', () {
    test('basically works', () async {
      var r = IoRestClient.root('https://ruian.fnx.io/');
      var rch = r.child(
          'api/v1/ruian/validate?apiKey=d9a14b87e627c3afea6a5b0ee6c836de065524de7e36f71acfe645c3e186928a&municipalityName=Fr%C3%BDdlant&zip=46401&cp=4001&street=Z%C3%A1meck%C3%A1');

      var rr = await rch.get();
      expect(rr.data, isMap);
      expect(rr.data, isNotEmpty);
      expect((rr.data as Map)['status'], isNotNull);
    });

    test('supports proxy', () async {
      final inner = HttpClient();
      inner.findProxy = (uri) => 'PROXY localhost:61999';

      var r = IoRestClient.root('https://ruian.fnx.io/', inner);
      var rch = r.child(
          'api/v1/ruian/validate?apiKey=d9a14b87e627c3afea6a5b0ee6c836de065524de7e36f71acfe645c3e186928a&municipalityName=Fr%C3%BDdlant&zip=46401&cp=4001&street=Z%C3%A1meck%C3%A1');

      try {
        // attempt to connect to crazy proxy will fail
        var rr = await rch.get();
      } catch (e) {
        expect(e.toString(), contains('SocketException'));
      }
    });
  });
}


test/rest_listing_test.dart
import 'package:fnx_rest/src/rest_client.dart';
import 'package:fnx_rest/src/rest_listing.dart';
import 'package:test/test.dart';

import 'simple_rest_client_test.mocks.dart';

void main() {
  group('SimpleListDriver', () {
    group('when unpacking data, should', () {
      var simpleRestListingDriver = SimpleListDriver();
      test('set hasNext to true when data were accepted', () {
        var data = simpleRestListingDriver.unpackData([1, 2, 3]);
        expect(data.hasNext, isTrue);
      });
      test('set hasNext to false when data is empty', () {
        var data = simpleRestListingDriver.unpackData([]);
        expect(data.hasNext, isFalse);
      });
      test('return the whole response', () {
        var data = simpleRestListingDriver.unpackData([1, 2, 3]);
        expect(data.data, equals([1, 2, 3]));
      });
    });
    group('when preparing client', () {
      test('include the page query parameter with given page', () {
        var dummyClient = RestClient(MockRestHttpClient(), null, '/');
        var driver = SimpleListDriver();
        var c = driver.prepareClient(dummyClient, 10);
        expect(c.params['page'], equals('10'));
      });
    });
  });

  group('ListResultDriver', () {
    group('when unpacking data, should', () {
      group('fail if', () {
        test('response is not a map', () {
          var driver = ListResultDriver();
          expect(
              () => driver.unpackData([1, 2, 3]),
              throwsA(
                  (String msg) => msg.contains('Expected result to be a map')));
        });
        test("response has no property 'data'", () {
          var driver = ListResultDriver();
          expect(
              () => driver.unpackData({'data': 'string'}),
              throwsA((String msg) =>
                  msg.contains("Expected 'data' key to be a List")));
        });
      });
    });
  });
}


example/main.dart
import 'package:fnx_rest/fnx_rest_io.dart';

void main() async {
  // Init your Rest API client
  RestClient apiRoot = IoRestClient.root(
      'https://jsonplaceholder.typicode.com'); // Use BrowserRestClient in browser ..

  // configure global headers
  apiRoot.setHeader('Authorization', 'FacelessMan');

  // follow serverside endpoints structure ...
  var apiUsers = apiRoot.child('/users');
  print(apiUsers.url);
  var rr = await apiUsers.get();
  if (!rr.success) rr.throwError();
  print(rr.data);

  // follow serverside endpoints structure ...
  var myApiUser = apiUsers.child('/1');
  print(myApiUser.url);
  rr = await myApiUser.get();
  if (!rr.success) rr.throwError();
  print(rr.data);

  // customize payload handling
  // var myApiUserPhoto = myApiUser.child("/photo");
  // myApiUserPhoto.acceptsBinary("image/png");
}


lib/fnx_rest_io.dart
// Copyright (c) 2016, Tomucha. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

///
/// Developer and Angular 2 friendly REST client.
/// Do this:
///
///       RestClient root = HttpRestClient.root("http://myapi.example/api/v1");
///
/// ... and start using it. See [RestClient] for more information.
///
library fnx_rest;

import 'dart:async';
import 'dart:io';

import 'package:http/http.dart';
import 'package:http/io_client.dart';
import 'package:http/src/response.dart';
import 'package:pedantic/pedantic.dart';

import 'src/rest_client.dart';

export 'src/rest_client.dart';
export 'src/rest_listing.dart';

class IoRestClient extends RestClient {
  IoRestClient.root(String url, [HttpClient? inner]) : this(null, url, inner);
  IoRestClient(RestClient? parent, String url, [HttpClient? inner])
      : super(IOHttpClient(inner), parent, url);
}

class IOHttpClient extends RestHttpClient {
  final IOClient _client;

  IOHttpClient([HttpClient? inner]) : _client = IOClient(inner);

  @override
  Future<Response> get(String url,
      {dynamic data, Map<String, String>? headers}) async {
    var request = _createRequest('GET', url, data, headers);
    return Response.fromStream(await _client.send(request));
  }

  @override
  Future<Response> delete(String url,
      {dynamic data, Map<String, String>? headers}) async {
    var request = _createRequest('DELETE', url, data, headers);
    return Response.fromStream(await _client.send(request));
  }

  @override
  Future<Response> streamedRequest(
      String method, String url, int length, Stream uploadStream,
      {Map<String, String>? headers}) async {
    StreamSubscription? subscription;
    try {
      var request = StreamedRequest(method, Uri.parse(url));
      request.contentLength = length;
      if (headers != null) {
        request.headers.addAll(headers);
      }
      subscription = uploadStream
          .listen(request.sink.add as void Function(dynamic)?, onDone: () {
        request.sink.close();
        subscription!.cancel();
      });
      return await Response.fromStream(await _client.send(request));
    } finally {
      unawaited(subscription?.cancel());
    }
  }

  @override
  Future<Response> post(String url, data, {Map<String, String>? headers}) {
    return _client.post(Uri.parse(url), headers: headers, body: data);
  }

  @override
  Future<Response> put(String url, data, {Map<String, String>? headers}) {
    return _client.put(Uri.parse(url), headers: headers, body: data);
  }

  @override
  Future<Response> head(String url, {Map<String, String>? headers}) {
    return _client.head(Uri.parse(url), headers: headers);
  }

  @override
  Future<Response> patch(String url, data, {Map<String, String>? headers}) {
    return _client.patch(Uri.parse(url), body: data, headers: headers);
  }

  Request _createRequest(
      String method, String url, dynamic data, Map<String, String>? headers) {
    var request = Request(method, Uri.parse(url));
    if (headers != null) {
      request.headers.addAll(headers);
    }
    if (data != null) {
      request.body = data;
    }
    return request;
  }
}


lib/src/debouncer.dart
import 'dart:async';

/// Debouncer is a Stream transformer which will wait for specified amount of time
/// before passing given value through the stream.
///
/// It can be useful for debouncing user events (scroll events, or button clicks).
///
/// If more values are delivered while the debouncer is waiting, only the last value
/// will be delivered and all other will be discarded.
///
/// Usage example:
///
///     var debouncedStream = someStream.transform(new Debouncer(new Duration(milliseconds: 200)))
///     debouncedStream.listen((data){
///       print("$data");
///     }
class Debouncer<S, T> implements StreamTransformer<S, T> {

  StreamController _controller;

  StreamSubscription _subscription;

  bool cancelOnError;

  // Original Stream
  Stream<S> _stream;

  Timer _timer;

  Duration _duration;

  Debouncer(Duration duration, {bool sync: false, this.cancelOnError}) {
    this._duration = duration;

    _controller = new StreamController<T>(onListen: _onListen,
        onCancel: _onCancel,
        onPause: () {
          _subscription.pause();
        },
        onResume: () {
          _subscription.resume();
        },
        sync: sync);
  }

  Duration get duration => _duration != null? _duration: new Duration(milliseconds: 100);

  void _onCancel() {
    _subscription.cancel();
    _subscription = null;
  }

  void _onListen() {
    _subscription = _stream.listen(onData,
        onError: _controller.addError,
        onDone: _controller.close,
        cancelOnError: cancelOnError);
  }

  void onData(S data) {
    _debounce(data);
  }

  void _debounce(S data) {
    if (_timer != null) _timer.cancel();
    _timer = new Timer(duration, () {
      if (_controller != null) _controller.add(data);
    });
  }

  @override
  Stream<T> bind(Stream<S> stream) {
    this._stream = stream;
    return _controller.stream;
  }
}


lib/fnx_rest_browser.dart
// Copyright (c) 2016, Tomucha. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

///
/// Developer and Angular 2 friendly REST client.
/// Do this:
///
///       RestClient root = HttpRestClient.root("http://myapi.example/api/v1");
///
/// ... and start using it. See [RestClient] for more information.
///
library fnx_rest;

import 'dart:async';

import 'package:http/browser_client.dart' as http;
import 'package:http/http.dart';
import 'package:pedantic/pedantic.dart';

import 'src/rest_client.dart';

export 'src/rest_client.dart';
export 'src/rest_listing.dart';

class BrowserRestClient extends RestClient {
  BrowserRestClient.root(String url) : this(null, url);
  BrowserRestClient(RestClient? parent, String url)
      : super(BrowserHttpClient(), parent, url);
}

class BrowserHttpClient extends RestHttpClient {
  final http.BrowserClient _client = http.BrowserClient();

  @override
  Future<Response> get(String url,
      {dynamic data, Map<String, String>? headers}) async {
    var request = _createRequest('GET', url, data, headers);
    return Response.fromStream(await _client.send(request));
  }

  @override
  Future<Response> delete(String url,
      {dynamic data, Map<String, String>? headers}) async {
    var request = _createRequest('DELETE', url, data, headers);
    return Response.fromStream(await _client.send(request));
  }

  @override
  Future<Response> post(String url, data, {Map<String, String>? headers}) {
    return _client.post(Uri.parse(url), headers: headers, body: data);
  }

  @override
  Future<Response> put(String url, data, {Map<String, String>? headers}) {
    return _client.put(Uri.parse(url), headers: headers, body: data);
  }

  @override
  Future<Response> head(String url, {Map<String, String>? headers}) {
    return _client.head(Uri.parse(url), headers: headers);
  }

  @override
  Future<Response> patch(String url, data, {Map<String, String>? headers}) {
    return _client.patch(Uri.parse(url), body: data, headers: headers);
  }

  @override
  Future<Response> streamedRequest(
      String method, String url, int length, Stream uploadStream,
      {Map<String, String>? headers}) async {
    StreamSubscription? subscription;
    try {
      var request = StreamedRequest(method, Uri.parse(url));
      if (headers != null) {
        request.headers.addAll(headers);
      }
      request.contentLength = length;
      subscription = uploadStream
          .listen(request.sink.add as void Function(dynamic)?, onDone: () {
        request.sink.close();
        subscription!.cancel();
      });
      return Response.fromStream(await _client.send(request));
    } finally {
      unawaited(subscription?.cancel());
    }
  }

  Request _createRequest(
      String method, String url, dynamic data, Map<String, String>? headers) {
    var request = Request(method, Uri.parse(url));
    if (headers != null) {
      request.headers.addAll(headers);
    }
    if (data != null) {
      request.body = data;
    }
    return request;
  }
}


lib/src/rest_client.dart
import 'dart:async';
import 'dart:collection';
import 'dart:convert';

import 'package:http/http.dart';
import 'package:http/src/response.dart';

///
/// Defines function which si able to take Dart objects and serialize them for upload (POST, PUT).
typedef Serializer = dynamic Function(
    dynamic payload, Map<String, String> requestHeaders);

///
/// Defines function which is able to take HTTP response and deserialize, possibly into a Dart object.
/// Should return a String or List<int> or binary.
typedef Deserializer = dynamic Function(Response response);

/// The rest client, this is the class you want to use.
/// On web use the [HttpRestClient.root] constructor to obtain preconfigured
/// RestClient.
///
class RestClient {
  late RestHttpClient _httpClient;

  RestClient? _parent;
  late String _url;

  Map<String, String> _headers = {};
  Map<String, dynamic /* String or List<String> */ > _params = {};

  Accepts? _accepts;
  Produces? _produces;
  int _workingCount = 0;

  final StreamController<bool> _workingStreamController =
      StreamController.broadcast();

  Stream<bool> get workingStream => _workingStreamController.stream;

  RestClient(RestHttpClient httpClient, RestClient? parent, String url,
      {Map<String, String>? headers}) {
    _parent = parent;
    var parsedUrl = parseUrl(url);
    _url = parsedUrl.url;
    _params = parsedUrl.params;
    _httpClient = httpClient;
    _headers = headers ?? HashMap<String, String>();
    if (parent == null) {
      accepts('application/json', defaultJsonDeserializer);
      produces('application/json', defaultJsonSerializer);
    }
  }

  RestClient child(String urlPart, {Map<String, String>? headers}) {
    return RestClient(_httpClient, this, urlPart, headers: headers);
  }

  /// Configure Accept header with appropriate [Deserializer].
  RestClient accepts(String mime, Deserializer deserializer) {
    _accepts = Accepts(mime, deserializer);

    return this;
  }

  /// Configure Accept header to receive binary data without any processing.
  RestClient acceptsBinary(String mime) {
    _accepts = Accepts(mime, defaultBinaryDeserializer);

    return this;
  }

  /// Configure Content-Type header to submit data as List<int> without serialization.
  RestClient producesBinary(String mime) {
    _produces = Produces(mime, defaultBinarySerializer);

    return this;
  }

  /// Configure Content-Type header with appropriate [Serializer].
  RestClient produces(String mime, Serializer serializer) {
    _produces = Produces(mime, serializer);

    return this;
  }

  /// HTTP get with optional additional headers and optional body.
  Future<RestResult> get({dynamic data, Map<String, String>? headers}) {
    var allHeaders = _headersToSend(headers ?? HashMap<String, String>());
    _addContentTypeHeader(allHeaders);
    _workStarted();
    var requestPayload = effProduces!.serialize(data, allHeaders);
    var resp = _httpClient.get(renderUrl(url, params),
        data: requestPayload, headers: allHeaders);
    return handleResponse(resp);
  }

  /// HTTP POST with optional additional headers. Data will be processed using configured [Serializer], JSON.encode(...) by default.
  Future<RestResult> post(dynamic data, {Map<String, String>? headers}) {
    var headersToSend = _headersToSend(headers ?? HashMap<String, String>());
    _addContentTypeHeader(headersToSend);
    _workStarted();
    var requestPayload = effProduces!.serialize(data, headersToSend);
    var resp = _httpClient.post(renderUrl(url, params), requestPayload,
        headers: headersToSend);
    return handleResponse(resp);
  }

  /// HTTP POST with optional additional headers. Data will be processed using configured [Serializer], JSON.encode(...) by default.
  Future<RestResult> patch(dynamic data, {Map<String, String>? headers}) {
    var headersToSend = _headersToSend(headers ?? HashMap<String, String>());
    _addContentTypeHeader(headersToSend);
    _workStarted();
    var requestPayload = effProduces!.serialize(data, headersToSend);
    var resp = _httpClient.patch(renderUrl(url, params), requestPayload,
        headers: headersToSend);
    return handleResponse(resp);
  }

  /// HTTP PUT with optional additional headers. Data will be processed using configured [Serializer], JSON.encode(...) by default.
  Future<RestResult> put(dynamic data, {Map<String, String>? headers}) {
    var headersToSend = _headersToSend(headers ?? HashMap<String, String>());
    _addContentTypeHeader(headersToSend);
    _workStarted();
    var requestPayload = effProduces!.serialize(data, headersToSend);
    var resp = _httpClient.put(renderUrl(url, params), requestPayload,
        headers: headersToSend);
    return handleResponse(resp);
  }

  /// HTTP DELETE with optional additional headers and optional body
  Future<RestResult> delete({dynamic data, Map<String, String>? headers}) {
    var allHeaders = _headersToSend(headers ?? HashMap<String, String>());
    _addContentTypeHeader(allHeaders);
    _workStarted();
    var requestPayload = effProduces!.serialize(data, allHeaders);
    var resp = _httpClient.delete(renderUrl(url, params),
        data: requestPayload, headers: allHeaders);
    return handleResponse(resp);
  }

  /// HTTP HEAD with optional additional headers.
  Future<RestResult> head({Map<String, String>? headers}) {
    var allHeaders = _headersToSend(headers ?? HashMap<String, String>());
    _workStarted();
    var resp = _httpClient.head(renderUrl(url, params), headers: allHeaders);
    return handleResponse(resp);
  }

  Future<RestResult> streamedRequest(
      String method, int contentLength, Stream uploadStream,
      {required Map<String, String> headers}) {
    var allHeaders = _headersToSend(headers);
    _workStarted();
    var resp = _httpClient.streamedRequest(
        method, renderUrl(url, params), contentLength, uploadStream,
        headers: allHeaders);
    return handleResponse(resp);
  }

  Map<String, String> _headersToSend(Map<String, String> headers) {
    var allHeaders = <String, String>{};
    allHeaders.addAll(this.headers);
    allHeaders.addAll(headers);
    _addAcceptHeader(allHeaders);
    return allHeaders;
  }

  Future<RestResult> handleResponse(Future<Response> resp) {
    return processResponse(effAccepts!, resp.whenComplete(_workCompleted));
  }

  static Future<RestResult> processResponse(
      Accepts accepts, Future<Response> resp) async {
    var r = await resp;
    dynamic data = accepts.deserialize(r);
    var result = RestResult(r.statusCode, data, r.headers);
    return result;
  }

  String get url => joinUrls(_parent?.url ?? '', _url);

  String get urlWithParams => renderUrl(url, params);

  static RegExp SLASH_CHAR = RegExp(r'\/');

  static UrlParseResult parseUrl(String url) {
    if (url.isEmpty) {
      return UrlParseResult('', {});
    } else {
      var parsed = Uri.parse(url);
      var parsedParams = <String, dynamic>{};
      // split the parameters from the url given
      var queryParameters = parsed.queryParameters;
      parsedParams.addAll(queryParameters);

      // we are only interested into plain url without query parameters (we store these separately)
      var plain = parsed.replace(queryParameters: {}, query: null);
      var plainUrl = plain.toString();
      if (plainUrl.endsWith('?')) {
        plainUrl = plainUrl.substring(0, plainUrl.length - 1);
      }
      return UrlParseResult(plainUrl, parsedParams);
    }
  }

  static String joinUrls(String base, String part) {
    if (base.isEmpty) return part;
    if (part.isEmpty) return base;

    var baseEndsWithSlash = base.endsWith('/');
    var partStartsWithSlash = part.startsWith(SLASH_CHAR);

    if (baseEndsWithSlash && partStartsWithSlash) {
      part = part.substring(1);
    } else if (!baseEndsWithSlash && !partStartsWithSlash) {
      part = '/$part';
    }

    return '$base$part';
  }

  static String renderUrl(String template, Map<String, dynamic> params) {
    var parsed = Uri.parse(template);
    var plainUrl = parsed.replace(queryParameters: params).toString();
    if (plainUrl.endsWith('?')) {
      plainUrl = plainUrl.substring(0, plainUrl.length - 1);
    }
    return plainUrl;
  }

  void _addAcceptHeader(Map<String, String> headers) {
    var mime = acceptsMime;
    if (mime == null) return;
    headers['Accept'] = mime;
  }

  void _addContentTypeHeader(Map<String, String> headers) {
    var mime = producesMime;
    if (mime == null) return;
    headers['Content-Type'] = mime;
  }

  Accepts? get effAccepts => _accepts ?? _parent?.effAccepts;
  Produces? get effProduces => _produces ?? _parent?.effProduces;

  String? get producesMime => effProduces?.mime;
  String? get acceptsMime => effAccepts?.mime;

  Serializer? get serializer => effProduces?.serializer;
  Deserializer? get deserializer => effAccepts?.deserializer;

  bool get working => _workingCount > 0;

  void _workStarted() {
    _workingCount++;
    _parent?._workStarted();
    _notifyWorkingState();
  }

  void _workCompleted() {
    _workingCount--;
    _parent?._workCompleted();
    _notifyWorkingState();
  }

  Map<String, String> get headers {
    var res = <String, String>{};

    var parentHeaders = _parent?.headers;
    if (parentHeaders != null && parentHeaders.isNotEmpty) {
      res.addAll(parentHeaders);
    }

    if (_headers.isNotEmpty) {
      res.addAll(_headers);
    }

    return res;
  }

  /// Configure additional header. Will be inherited by all children.
  RestClient setHeader(String name, String value) {
    _headers[name] = value;
    return this;
  }

  /// Remove additional header. Will be inherited by all children.
  RestClient removeHeader(String name) {
    _headers.remove(name);
    return this;
  }

  Map<String, dynamic /* String or List<String>*/ > get params {
    var res = <String, dynamic>{};
    if (_parent != null) {
      res.addAll(_parent!.params);
    }
    res.addAll(_params);
    return res;
  }

  /// Add query parameter.
  RestClient setParam(String name, String value) {
    return _setParam(name, value);
  }

  /// Add query parameter.
  RestClient setParams(String name, List<String> value) {
    return _setParam(name, value);
  }

  String? getParam(String name) {
    dynamic res = _getParam(name);
    if (res is String) {
      return res;
    } else if (res == null) {
      return null;
    } else {
      throw 'Invalid parameter type! \"$name\" should be of type String. Was $res';
    }
  }

  List<String>? getParams(String name) {
    dynamic res = _getParam(name);

    if (res is String) {
      return [res];
    } else if (res is List<String>) {
      var all = <String>[];
      all.addAll(res);
      return all;
    } else if (res == null) {
      return null;
    } else {
      throw 'Invalid parameter type! Only String an List<String> are supported $res';
    }
  }

  dynamic _getParam(String name) {
    return params[name];
  }

  RestClient _setParam(String name, /* String or List<String>*/ dynamic value) {
    if (value == null) {
      _params.remove(name);
    } else {
      _params[name] = value;
    }

    return this;
  }

  void _notifyWorkingState() {
    _workingStreamController.add(working);
  }
}

///
/// Create your own implementation of this class if you
/// need to inject your own HTTP client into RestClient.
/// On server, for example. See [RestClient].
///
abstract class RestHttpClient {
  Future<Response> get(String url,
      {dynamic data, Map<String, String>? headers});
  Future<Response> post(String url, dynamic data,
      {Map<String, String>? headers});
  Future<Response> put(String url, dynamic data,
      {Map<String, String>? headers});
  Future<Response> delete(String url,
      {dynamic data, Map<String, String>? headers});
  Future<Response> head(String url, {Map<String, String>? headers});
  Future<Response> patch(String url, dynamic data,
      {Map<String, String> headers});
  Future<Response> streamedRequest(
      String method, String url, int length, Stream uploadStream,
      {Map<String, String>? headers});
}

RegExp microRemoval = RegExp(r'\.[0-9]{0,6}');

Object? toJsonEncodable(Object? value) {
  if (value is DateTime) {
    return value.toUtc().toIso8601String().replaceAll(microRemoval, r'');
  } else {
    return value;
  }
}

///
/// Uses JSON.decode(...) from dart:convert.
Deserializer defaultJsonDeserializer = (Response response) {
  final payload = response.body;
  if (payload is String) {
    if (payload.trim().isEmpty) {
      return null;
    } else {
      return json.decode(payload);
    }
  } else {
    throw RestClientException('Payload should be string if parsed as JSON');
  }
};

///
/// Uses JSON.encode(...) from dart:convert.
Serializer defaultJsonSerializer =
    (dynamic payload, Map<String, String> requestHeaders) {
  if (payload == null || (payload is String && payload.trim().isEmpty)) {
    return null;
  } else {
    return json.encode(payload, toEncodable: toJsonEncodable);
  }
};

Deserializer defaultBinaryDeserializer =
    (Response response) => response.bodyBytes;

Serializer defaultBinarySerializer =
    (dynamic payload, Map<String, String> requestHeaders) => payload;

class RestClientException implements Exception {
  final String message;

  RestClientException(this.message);
}

///
/// Result of REST call.
class RestResult {
  /// HTTP status
  final int status;

  /// Deserialized response body.
  final dynamic data;

  final Map<String, String> headers;

  RestResult(this.status, this.data, this.headers);

  /// Convenient interpretation of HTTP status (>=200 && < 300)
  bool get success => 200 <= status && status < 300;

  /// Convenient interpretation of HTTP status (>=400 && < 500)
  bool get failure => 400 <= status && status < 500;

  /// Convenient interpretation of HTTP status (>=500)
  bool get error => 500 <= status;

  /// Convenient interpretation of HTTP status (404)
  bool get notFound => status == 404;

  /// Convenient interpretation of HTTP status (401)
  bool get notAuthorized => status == 401;

  /// Convenient interpretation of HTTP status (403)
  bool get forbidden => status == 403;

  /// If the result is success method returns data. If it's not, it throws 'this'.
  dynamic get successData {
    assertSuccess();
    return data;
  }

  void assertSuccess() {
    if (!success) throwError();
  }

  void throwError() {
    throw HttpException(status, data);
  }
}

class HttpException {
  int httpStatus;
  var data;

  HttpException(this.httpStatus, [this.data]);

  @override
  String toString() {
    if (data == null)
      return 'HttpException{httpStatus: $httpStatus, data: $data}';
    if ((data is String) && data.length < 1000)
      return 'HttpException{httpStatus: $httpStatus, data: $data}';
    return 'HttpException{httpStatus: $httpStatus}';
  }
}

class UrlParseResult {
  String url;
  Map<String, dynamic> _params = {};

  UrlParseResult(this.url, Map<String, dynamic> argParams) {
    _params = argParams;
  }

  Map<String, dynamic> get params => _params;
}

///
/// Couple of mime type and Deserializer.
///
class Accepts {
  final String mime;
  final Deserializer deserializer;

  Accepts(this.mime, this.deserializer);

  dynamic deserialize(Response resp) => deserializer(resp);
}

///
/// Couple of mime type and Serializer.
///
class Produces {
  final String mime;
  final Serializer serializer;

  Produces(this.mime, this.serializer);

  dynamic serialize(dynamic payload, Map<String, String> requestHeaders) =>
      serializer(payload, requestHeaders);
}
